### PR TITLE
Am/feat/parallel pks

### PR DESCRIPTION
**Pull request by IceTDrinker**
refs: https://github.com/zama-ai/tfhe-rs-internal/issues/71

### PR content/description

propose a parallel implementation for the packing keyswitch

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs [hunk 1]
```diff
@@ -376,3 +378,343 @@ pub fn keyswitch_lwe_ciphertext_list_and_pack_in_glwe_ciphertext<
...
+/// );
+///
+/// let mut output_glwe = GlweCiphertext::new(
+///     0u64,
+///     output_glwe_secret_key.glwe_dimension().to_glwe_size(),
+///     output_glwe_secret_key.polynomial_size(),
+///     ciphertext_modulus,
+/// );
+///
+/// par_keyswitch_lwe_ciphertext_into_glwe_ciphertext(&pksk, &input_lwe, &mut output_glwe);
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs:448]: wrong function called 🙄

File: tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs [hunk 2]
```diff
@@ -376,3 +378,343 @@ pub fn keyswitch_lwe_ciphertext_list_and_pack_in_glwe_ciphertext<
...
+/// let mut output_glwe = GlweCiphertext::new(
+///     0u64,
+///     output_glwe_secret_key.glwe_dimension().to_glwe_size(),
+///     output_glwe_secret_key.polynomial_size(),
+///     ciphertext_modulus,
+/// );
+///
+/// // Try to use 4 threads for the keyswitch if enough are available
+/// // in the current rayon thread pool
+/// par_keyswitch_lwe_ciphertext_into_glwe_ciphertext_with_thead_count(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs:565]: same thing

File: tfhe/src/core_crypto/algorithms/lwe_keyswitch.rs [hunk 3]
```diff
@@ -417,12 +417,12 @@ pub fn par_keyswitch_lwe_ciphertext_with_thread_count<Scalar, KSKCont, InputCont
         lwe_keyswitch_key.decomposition_level_count(),
     );
 
-    let mut intermediate_accumulators: Vec<_> = Vec::with_capacity(thread_count.0);
+    // Don't go above the current number of threads
+    let thread_count = thread_count.0.min(rayon::current_num_threads());
+    let mut intermediate_accumulators: Vec<_> = Vec::with_capacity(thread_count);
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/lwe_keyswitch.rs:422]: nitpick: the `: Vec<_>` anotation should be removed as the function used is `Vec::` and the annotation does not precise the type.

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/lwe_keyswitch.rs:422]: indeed, that looks silly

File: tfhe/src/core_crypto/algorithms/lwe_private_functional_packing_keyswitch.rs [hunk 4]
```diff
@@ -162,6 +162,11 @@ pub fn par_private_functional_keyswitch_lwe_ciphertext_into_glwe_ciphertext_with
         "This operation currently only supports native moduli"
     );
 
+    assert!(
+        thread_count.0 != 0,
+        "Got thread_count == 0, this is not supported"
+    );
+
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/lwe_private_functional_packing_keyswitch.rs:169]: Clippy did not complain about using `assert!` and not `assert_ne!` ? 😮

File: tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs [hunk 5]
```diff
@@ -376,3 +378,378 @@ pub fn keyswitch_lwe_ciphertext_list_and_pack_in_glwe_ciphertext<
...
+                });
+
+            buffer
+        })
+        .collect_into_vec(&mut intermediate_buffers);
+
+    let result = intermediate_buffers
+        .par_iter_mut()
+        .reduce_with(|lhs, rhs| {
+            slice_wrapping_add_assign(lhs.as_mut(), rhs.as_mut());
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs:747]: should it be `rhs.as_ref()` ?

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs:747]: you are right

File: tfhe/src/core_crypto/algorithms/lwe_private_functional_packing_keyswitch.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/lwe_private_functional_packing_keyswitch.rs:169]: probably because I added an error message

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**